packages/core/database/lib/entity-manager/relations-orderer.js

```diff
@@ -1,6 +1,106 @@
 'use strict';
 
 const _ = require('lodash/fp');
+const { InvalidRelationError } = require('../errors');
+/**
+ * When connecting relations, the order you connect them matters.
+ *
+ * Example, if you connect the following relations:
+ *   { id: 5, position: { before: 1 } }
+ *   { id: 1, position: { before: 2 } }
+ *   { id: 2, position: { end: true } }
+ *
+ * Going through the connect array, id 5 has to be connected before id 1,
+ * so the order of id5 = id1 - 1. But the order value of id 1 is unknown.
+ * The only way to know the order of id 1 is to connect it first.
+ *
+ * This function makes sure the relations are connected in the right order:
+ *   { id: 2, position: { end: true } }
+ *   { id: 1, position: { before: 2 } }
+ *   { id: 5, position: { before: 1 } }
+ *
+ */
+const sortConnectArray = (connectArr, initialArr = [], strictSort = true) => {
+  const sortedConnect = [];
+  // Boolean to know if we have to recalculate the order of the relations
+  let needsSorting = false;
+  // Map to validate if relation is already in sortedConnect or DB.
+  const relInArray = initialArr.reduce((acc, rel) => ({ ...acc, [rel.id]: true }), {});
+  // Map to store the first index where a relation id is connected
+  const firstSeen = {};
+  // Map to validate if connect relation has already been computed
+  const computedIdx = {};
+
+  connectArr.forEach((rel, idx) => {
+    // If adjacent relation is not in the database or seen yet in the connect array
+    // then we need to sort the connect array
+    const adjacentRelId = rel.position?.before || rel.position?.after;
+    if (!relInArray[adjacentRelId] && !firstSeen[adjacentRelId]) needsSorting = true;
+    // Populate firstSeen
+    if (!(rel.id in firstSeen)) firstSeen[rel.id] = idx;
+  });
```

Comment:
I think you can do a reduce here to map your "firstSeen" object.
```suggestion
  const mappedRelationsIndex = connectArr.reduce((mapper, relation, index) => {
    // If adjacent relation is not in the database or seen yet in the connect array
    // then we need to sort the connect array
    const adjacentRelId = relation.position?.before || relation.position?.after;

    // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Logical_OR_assignment
    needsSorting ||= !relInArray[adjacentRelId] && !mapper[adjacentRelId];

     return {
       [relation.id]: index,
       ...mapper,
     };
  }, {});
```

I have not tested this function, so there may be some small errors (!)

Reply:
sounds better, I will go this way.
Btw the logical or assignment is only compatible from node v15, and we support node v14 so I will use the old boring way :/

Reply:
Oh, thought it was on Node@14, my bad 😞